src/execution_error.rs

```diff
@@ -0,0 +1,275 @@
+    where
+        F: FnOnce() -> Context,
+    {
+        self.map_err(|e| match e {
+            cancelled @ ExecutionError::Cancelled(_) => cancelled,
+            _ => ExecutionError::InContext(with_context(), Box::new(e)),
+        })
+    }
+}
+
+impl ExecutionError {
+    pub fn display_pretty<'a>(
+        &'a self,
+        source_path: &'a Path,
+        source: &'a str,
+        tsg_path: &'a Path,
+        tsg: &'a str,
+    ) -> impl std::fmt::Display + 'a {
+        DisplayExecutionErrorPretty {
+            error: self,
+            source_path,
+            source,
+            tsg_path,
+            tsg,
+        }
+    }
+}
+
+struct DisplayExecutionErrorPretty<'a> {
+    error: &'a ExecutionError,
+    source_path: &'a Path,
+    source: &'a str,
+    tsg_path: &'a Path,
+    tsg: &'a str,
+}
+
+impl std::fmt::Display for DisplayExecutionErrorPretty<'_> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.fmt_entry(f, 0, self.error)
+    }
+}
+
+impl DisplayExecutionErrorPretty<'_> {
+    fn fmt_entry(
+        &self,
+        f: &mut std::fmt::Formatter<'_>,
+        index: usize,
+        error: &ExecutionError,
+    ) -> std::fmt::Result {
+        match error {
+            ExecutionError::InContext(context, cause) => {
+                match context {
+                    Context::Statement {
+                        statement,
+                        statement_location,
+                        stanza_location,
+                        source_location,
+                        node_kind,
+                    } => {
+                        writeln!(f, "{:>5}: Error executing statement {}", index, statement)?;
+                        write!(
+                            f,
+                            "{}",
+                            Excerpt::from_source(self.tsg_path, self.tsg, statement_location, 7)
+                        )?;
+                        writeln!(f, "{}in stanza", " ".repeat(7))?;
+                        write!(
+                            f,
+                            "{}",
+                            Excerpt::from_source(self.tsg_path, self.tsg, stanza_location, 7)
+                        )?;
+                        writeln!(f, "{}matching ({}) node", " ".repeat(7), node_kind)?;
+                        write!(
+                            f,
+                            "{}",
+                            Excerpt::from_source(self.source_path, self.source, source_location, 7)
+                        )?;
+                        Ok(())
+                    }
+                    Context::Other(msg) => writeln!(f, "{:>5}: {}", index, msg),
+                }?;
+                self.fmt_entry(f, index + 1, cause)
+            }
+            other => writeln!(f, "{:>5}: {}", index, other),
+        }
+    }
+}
+
+/// Excerpts of source from either the target language file or the tsg rules file.
+pub struct Excerpt<'a> {
```

Comment:
I realize this isn't introducing this code, just moving it, but we might consider using the [`miette`](https://docs.rs/miette/*/miette/) crate for constructing "nested context based on source code".  It has a very nice [data model](https://docs.rs/miette/5.3.0/miette/#-snippets) for these kinds of errors, and also has a good ability to separate error construction from error display/reporting.

Reply:
I have a déjà-vu. I feel like this was discussed before, and my reply is the same as last time 😛: If we want this, let's do it in another PR.

Reply:
Agreed